File: samples/CameraAccess/CameraAccess/Grok/WakeWordManager.swift
import Foundation
import Porcupine

enum WakeWordState: Equatable {
  case disabled
  case notConfigured
  case idle
  case listening
  case detected
  case error(String)

  var displayText: String {
    switch self {
    case .disabled: return "Wake Off"
    case .notConfigured: return "Wake Setup"
    case .idle: return "Wake"
    case .listening: return "Listening"
    case .detected: return "Wake Detected"
    case .error: return "Wake Error"
    }
  }
}

@MainActor
final class WakeWordManager {
  private var porcupineManager: PorcupineManager?

  private(set) var state: WakeWordState = .disabled
  var onStateChanged: ((WakeWordState) -> Void)?
  var onDetected: (() -> Void)?

  func startListening() {
    guard SettingsManager.shared.wakeWordEnabled else {
      updateState(.disabled)
      return
    }

    guard SettingsManager.shared.isPorcupineConfigured else {
      updateState(.notConfigured)
      return
    }

    stopListening(updateIdleState: false)

    do {
      let accessKey = SettingsManager.shared.porcupineAccessKey
      let sensitivity = Float32(SettingsManager.shared.wakeWordSensitivity)
      let detection: (Int32) -> Void = { [weak self] _ in
        Task { @MainActor in
          guard let self else { return }
          self.updateState(.detected)
          self.onDetected?()
        }
      }
      let errorCallback: (Error) -> Void = { [weak self] error in
        Task { @MainActor in
          self?.updateState(.error(error.localizedDescription))
        }
      }

      if let keywordPath = resolvedCustomKeywordPath() {
        porcupineManager = try PorcupineManager(
          accessKey: accessKey,
          keywordPath: keywordPath,
          sensitivity: sensitivity,
          onDetection: detection,
          errorCallback: errorCallback
        )
      } else {
        porcupineManager = try PorcupineManager(
          accessKey: accessKey,
          keyword: builtInKeyword(named: SettingsManager.shared.wakeWordBuiltInKeyword),
          sensitivity: sensitivity,
          onDetection: detection,
          errorCallback: errorCallback
        )
      }

      try porcupineManager?.start()
      updateState(.listening)
    } catch {
      updateState(.error(error.localizedDescription))
      tearDownManager()
    }
  }

  func stopListening(updateIdleState: Bool = true) {
    tearDownManager()
    if updateIdleState {
      updateState(SettingsManager.shared.wakeWordEnabled ? .idle : .disabled)
    }
  }

  private func tearDownManager() {
    if let manager = porcupineManager {
      try? manager.stop()
      try? manager.delete()
    }
    porcupineManager = nil
  }

  private func resolvedCustomKeywordPath() -> String? {
    let rawPath = SettingsManager.shared.wakeWordKeywordPath.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !rawPath.isEmpty else { return nil }

    if FileManager.default.fileExists(atPath: rawPath) {
      return rawPath
    }

    let url = URL(fileURLWithPath: rawPath)
    let baseName = url.deletingPathExtension().lastPathComponent
    let ext = url.pathExtension.isEmpty ? "ppn" : url.pathExtension
    return Bundle.main.path(forResource: baseName, ofType: ext)
  }

  private func builtInKeyword(named name: String) -> Porcupine.BuiltInKeyword {
    switch name.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() {
    case "alexa": return .alexa
    case "americano": return .americano
    case "blueberry": return .blueberry
    case "bumblebee": return .bumblebee
    case "computer": return .computer
    case "grapefruit": return .grapefruit
    case "grasshopper": return .grasshopper
    case "hey google": return .heyGoogle
    case "hey siri": return .heySiri
    case "ok google": return .okGoogle
    case "picovoice": return .picovoice
    case "porcupine": return .porcupine
    case "terminator": return .terminator
    case "jarvis": fallthrough
    default: return .jarvis
    }
  }

  private func updateState(_ newState: WakeWordState) {
    state = newState
    onStateChanged?(newState)
  }
}

File: samples/CameraAccessAndroid/app/src/main/java/com/meta/wearable/dat/externalsampleapps/cameraaccess/grok/WakeWordManager.kt
package com.meta.wearable.dat.externalsampleapps.cameraaccess.grok

import ai.picovoice.porcupine.Porcupine
import ai.picovoice.porcupine.PorcupineManager
import android.content.Context
import android.util.Log
import com.meta.wearable.dat.externalsampleapps.cameraaccess.settings.SettingsManager
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

sealed class WakeWordState(val displayText: String) {
    data object Disabled : WakeWordState("Wake Off")
    data object NotConfigured : WakeWordState("Wake Setup")
    data object Idle : WakeWordState("Wake")
    data object Listening : WakeWordState("Listening")
    data object Detected : WakeWordState("Wake Detected")
    data class Error(val message: String) : WakeWordState("Wake Error")
}

class WakeWordManager(
    private val context: Context,
    private val scope: CoroutineScope,
    private val onDetected: () -> Unit,
) {
    companion object {
        private const val TAG = "WakeWordManager"
    }

    private var porcupineManager: PorcupineManager? = null
    private val _state = MutableStateFlow<WakeWordState>(WakeWordState.Disabled)
    val state: StateFlow<WakeWordState> = _state.asStateFlow()

    fun startListening() {
        if (!SettingsManager.wakeWordEnabled) {
            _state.value = WakeWordState.Disabled
            return
        }

        if (!SettingsManager.isPorcupineConfigured) {
            _state.value = WakeWordState.NotConfigured
            return
        }

        stopListening(updateIdleState = false)

        try {
            var builder = PorcupineManager.Builder()
                .setAccessKey(SettingsManager.porcupineAccessKey)
                .setSensitivity(SettingsManager.wakeWordSensitivity)
                .setErrorCallback { error ->
                    scope.launch {
                        _state.value = WakeWordState.Error(error.message ?: error.toString())
                    }
                }

            val customKeywordPath = SettingsManager.wakeWordKeywordPath.trim()
            builder = if (customKeywordPath.isNotEmpty()) {
                builder.setKeywordPath(customKeywordPath)
            } else {
                builder.setKeyword(builtInKeyword(SettingsManager.wakeWordBuiltInKeyword))
            }

            porcupineManager = builder.build(context.applicationContext) { _ ->
                scope.launch {
                    _state.value = WakeWordState.Detected
                    onDetected()
                }
            }
            porcupineManager?.start()
            _state.value = WakeWordState.Listening
        } catch (error: Exception) {
            Log.e(TAG, "Failed to start wake word listener", error)
            _state.value = WakeWordState.Error(error.message ?: error.toString())
            tearDownManager()
        }
    }

    fun stopListening(updateIdleState: Boolean = true) {
        tearDownManager()
        if (updateIdleState) {
            _state.value = if (SettingsManager.wakeWordEnabled) WakeWordState.Idle else WakeWordState.Disabled
        }
    }

    private fun tearDownManager() {
        porcupineManager?.let { manager ->
            try {
                manager.stop()
            } catch (error: Exception) {
                Log.d(TAG, "Wake word listener was not active", error)
            }
            manager.delete()
        }
        porcupineManager = null
    }

    private fun builtInKeyword(name: String): Porcupine.BuiltInKeyword =
        when (name.trim().lowercase().replace(" ", "_").replace("-", "_")) {
            "alexa" -> Porcupine.BuiltInKeyword.ALEXA
            "americano" -> Porcupine.BuiltInKeyword.AMERICANO
            "blueberry" -> Porcupine.BuiltInKeyword.BLUEBERRY
            "bumblebee" -> Porcupine.BuiltInKeyword.BUMBLEBEE
            "computer" -> Porcupine.BuiltInKeyword.COMPUTER
            "grapefruit" -> Porcupine.BuiltInKeyword.GRAPEFRUIT
            "grasshopper" -> Porcupine.BuiltInKeyword.GRASSHOPPER
            "hey_google" -> Porcupine.BuiltInKeyword.HEY_GOOGLE
            "hey_siri" -> Porcupine.BuiltInKeyword.HEY_SIRI
            "ok_google" -> Porcupine.BuiltInKeyword.OK_GOOGLE
            "picovoice" -> Porcupine.BuiltInKeyword.PICOVOICE
            "porcupine" -> Porcupine.BuiltInKeyword.PORCUPINE
            "terminator" -> Porcupine.BuiltInKeyword.TERMINATOR
            else -> Porcupine.BuiltInKeyword.JARVIS
        }
}
